[IDE] Refactoring: conversion strings concatenation to interpolation …

Kacper20 opened the pull request:
Initial implementation of conversion strings concatenation expression to interpolation refactor.
I would appreciate any feedback. It's my first touch to the AST & Refactoring Engine and there may be an easier way to do certain things.

I've written tests, but I'm in the middle of running it(had some problems with running through Xcode), so it may still fail. Working on it :)

Addresses [SR-5745](https://bugs.swift.org/browse/SR-5745).

cc: @nkcsgexi

Kacper20 committed:
[IDE] SR-5745 Refactoring: conversion strings concatenation expression to strings interpolation

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD->getName().str() != "+") {
```
We should compare decl pointer instead of its name for "+" here. Could you try to add a function like "getStringDecl()" in ASTContext and use it here?

Kacper20 replied:
Hmm... Could you help me a bit on this? I've changed the other string comparisons but this one is a bit more tricky. What I've done is: I've found `KnownDecls.def` which seems to define functions and is a potential place to dig in. I could add a new entry and as I see there is a macro that generated concrete implementation of the getter. But how should I name it(string literal form)? How does binding look like between this declaration and the object in memory? I haven't found any examples for using `FUNCDECL` macro with operators like `+`. Thanks in advance!

nkcsgexi replied:
I think we can use the function I added in #11964. Please let me know if that works.

Kacper20 replied:
Thanks for this! Unfortunately `FD` and `ASTContext getPlusFunctionOnRangeReplaceableCollection()` are not the same. That's weird since String from Swift 4 adopts this protocol. I'm using `FD` like in the example above. Should I maybe refer to it in some other way?

Kacper20 replied:
Hm, it's defined directly in `String.swift` here: https://github.com/apple/swift/blob/master/stdlib/public/core/String.swift#L873 It's probably why it doesn't work.

Kacper20 replied:
I've ended up adding new function on `ASTContext`

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+      if (E->getKind() == ExprKind::Binary) {
+        auto *BE = dyn_cast<BinaryExpr>(E);
+        auto *OperatorDeclRef = BE->getSemanticFn()->getMemberOperatorRef();
+        if (!isConcatenationExpr(OperatorDeclRef) || ExprType != "String") {
```
Same here. We should use `ExprType->getNominalOrBoundGenericNominal() == Ctx.getStringDecl()` here instead of comparing strings.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+        return true;
+      }
+      //Everything that evaluates to string should be gathered
+      if (ExprType == "String") {
```
Same here as above.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+        return false;
+      }
+      if (E->getKind() == ExprKind::DeclRef) {
+        auto *DR = dyn_cast<DeclRefExpr>(E);
```
The above two lines can be combined to `if (auto *DR = dyn_cast<DeclRefExpr>(E)) {`. Also, could you add a comment above why we need to handle this case specifically?

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+  //There should be two or more expressions to convert
+  if (!Walker.IsValidInterpolation || Walker.Bucket->size() < 2)
+    return nullptr;
+  return std::move(Walker.Bucket);
```
We don't need `std::move` here.

Kacper20 replied:
Hm, I'm returning `std::unique_ptr` which has deleted copy constructor, so I need to move the membership if the func signature is `static std::unique_ptr<llvm::SetVector<Expr*>>
  findConcatenatedExpressions(ResolvedRangeInfo Info)`. Am I missing something?

nkcsgexi replied:
oh, you are right. We don't need `std::move` if the unique pointer is a local variable.

Kacper20 replied:
Great, thanks!

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+  return std::move(Walker.Bucket);
+}
+
+StringRef static interpolatedExpressionForm(Expr* E, SourceManager& SM) {
```
`static` before `StringRef`.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+    return Literal->getValue();
+  }
+  auto CharRange = Lexer::getCharSourceRangeFromSourceRange(SM, E->getSourceRange());
+  auto ExpStr = SM.extractText(CharRange).str();
```
The above two lines can be combined as 
```
auto ExpStr = Lexer::getCharSourceRangeFromSourceRange(SM, E->getSourceRange()).str();
```

nkcsgexi reviewed (changes requested):
Great work so far! Some comments inline.

Kacper20 commented:
Thanks for the review! Working on fixes right now :)

xwu commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+        return true;
+
+      auto ExprType = E->getType().getString();
+      //Only binary concatenation operators should exist in expression
```
Drive-by nit: comments should have a space after `//` and end in punctuation.

Kacper20 commented:
Hi @nkcsgexi,

I've fixed all of the code according to the comments - thanks for them! I've implemented new function on `ASTContext` to get the + function declaration for Strings. Many thanks for showing me the way, it was really helpful and I learned a lot.

Two code transformation test cases are passing now. `RefactoringKind/basic.swift` test is failing but it's not in the code I've written: `118:8: error: expected '{' in body of function declaration`. Should this be like that, or should I fix it?

Please let me know whether I can improve something.

nkcsgexi commented on lib/AST/ASTContext.cpp:
```diff
@@ -571,6 +574,38 @@ FuncDecl *ASTContext::getPlusFunctionOnRangeReplaceableCollection() const {
+      auto SecondParamType =
+      SecondParam->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto StringDecl = getStringDecl();
+      if (FirstParamType == StringDecl && SecondParamType == StringDecl) {
```
Nitpick: Can we simplify the above logic by introducing a local closure `isStringParam` and use the closure on both `ParamLsts[1]->get[0]` and `ParamLsts[1]->get[1]`?

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,106 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD != Ctx.getPlusFunctionOnString()) {
```
Theoretically, both this '+' and the '+' function I added in the other commit could apply to strings. It's entirely up to type checker to decide the preferred one. Could you check both situations here?

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1448,6 +1448,106 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+  return std::move(Walker.Bucket);
+}
+
+static std::string interpolatedExpressionForm(Expr* E, SourceManager& SM) {
```
Instead of returning `std::string`, could we take a stream operator and output to the stream instead?
e.g.
```
  llvm::SmallString<64> Buffer;
  llvm::raw_svector_ostream OS(Buffer);
  interpolatedExpressionForm(E, SM, OS); // use OS << "something"
```

nkcsgexi reviewed (changes requested):
Great improvement! More comments inline.

Kacper20 committed:
Review fixes - more concise code in getStringFunc implementation. Working on streams instead of std::string directly.

Kacper20 commented:
Thanks, @nkcsgexi :)
Please let me know whether it's what you've expected.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@Kacper20 mind taking a look at these failing tests?

Kacper20 commented:
@nkcsgexi Yes, of course. Looking at it :)

Kacper20 commented:
@nkcsgexi I wonder one more thing: From where should I take the `ASTContext` in `isApplicable` call? I see that I could get it through `Info.RangeContext` and `Info.getType`. What's the preferred way to do this thing?

Kacper20 committed:
Fixed refactoring/ tests failing

Kacper20 commented:
@nkcsgexi I've fixed `refactoring/` tests but SourceKit ones are still failing. They're failing even if I'm commenting out the whole implementation, rebuilding and running tests again. Any ideas what might be wrong?

Kacper20 commented:
Tests are passing for me locally right now. There was weird error with SourceKit tests regarding calling pure virtual function but it doesn't occur now for me. Could we check whether this works on CI?

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

Kacper20 commented:
@CodaFi Could you please run again? A dumb mistake from my side done while hunting those SourceKit tests crashes ;). Thanks in advance!

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@Kacper20 I think getting it from `Info.RangeContext` makes perfect sense, something like `Info.RangeContext->getASTContext()`

Kacper20 commented:
@nkcsgexi Any idea why build on Linux fails?

harlanhaskins commented on include/swift/IDE/RefactoringKinds.def:
```diff
@@ -46,6 +46,7 @@ RANGE_REFACTORING(ExtractFunction, "Extract Method", extract.function)
 
 RANGE_REFACTORING(ExtractRepeatedExpr, "Extract Repeated Expression", extract.expr.repeated)
 
+RANGE_REFACTORING(ConvertStringsConcatenationToInterpolation, "Convert to Strings Interpolation", convert.string-concatenation.interpolation)
```
Nit: This should be `Convert to String Interpolation`

Kacper20 replied:
Thanks, fixed!

Kacper20 committed:
Fixed refactoring user visible name. Fixed warning of unused variable.

nkcsgexi commented on lib/IDE/Refactoring.cpp:
```diff
@@ -1555,6 +1555,116 @@ bool RefactoringActionCollapseNestedIfExpr::performChange() {
+}
+
+static void interpolatedExpressionForm(Expr* E, SourceManager& SM,
+                                              llvm::raw_svector_ostream& OS) {
```
Tiny nit: could we change the second parameter to type `llvm::raw_ostream &OS`?

Kacper20 replied:
Done. Fixed in all other places, too. Thanks!

nkcsgexi replied:
Thanks for fixing the name convention! I mean the type of the second parameter here should be `llvm:: raw_ostream` instead of `llvm::raw_svector_ostream` to be more acceptive in the future.

Kacper20 replied:
Oh, sorry :) Right now should be fine.

Kacper20 committed:
Fixed the arguments namings

nkcsgexi commented:
@swift-ci please smoke test

Kacper20 committed:
Changed type of stream to be more general

nkcsgexi commented:
@swift-ci Please clean smoke test

nkcsgexi commented:
@harlanhaskins any more comments?

harlanhaskins commented:
Nothing more from me! Great work, @Kacper20!

Kacper20 commented:
Thanks for all of the help in my first contribution to Swift OS @nkcsgexi, @harlanhaskins :)
